File: corpus/runs/openai-agents-js/exercise.mjs
/**
 * Exercises the pinned customer service example so an OpenAI Agents run in JavaScript can be joined to the
 * declared graph.
 *
 * The OpenAI Agents dialect had been measured in one ecosystem, and it is the richest source of defects this
 * build has had: `openai-cs-agents-demo-exercised` is where a handoff arriving as a tool call, an
 * instrumentation's own wrapper spans becoming components, and a provider written under only one of two
 * attribute names were each found. Every one of those is a claim about what one program emits.
 * `@arizeai/openinference-instrumentation-openai-agents` is a different program, and nothing in this build had
 * ever read a span it wrote.
 *
 * The example is the repository's own, and so are the agents, the tools and the handoffs between them. What
 * the driver supplies is the model, and three things follow from that.
 *
 * **The run is hermetic and needs no credential.** `setDefaultModelProvider` is the SDK's own way to say
 * where completions come from, so a model that answers from a script reaches nothing outside this process.
 *
 * **The run is reproducible, which a run against a provider was not.** Driven by `gpt-4o-mini` the same
 * conversation produced eight spans, then eleven, then eight: the seat agent follows a routine that asks for a
 * confirmation number, and whether it asks again before calling its tool is the model's choice. A corpus entry
 * pins a span count, so the decisions have to be the driver's.
 *
 * **The script chooses from what the request offers rather than from a name written here.** Each call takes
 * the first of the handoff and the tool that the agent actually holds and has not used, so what runs is
 * whatever the repository declared, and a rename there changes the run rather than silently skipping it.
 *
 * Two mechanics are worth stating because each one was a failure before it was a line of this file.
 *
 * **The example is a terminal loop.** It builds a readline interface over `process.stdin` at module scope and
 * asks for a message until the stream ends, so one scripted turn followed by end of file is the whole
 * conversation.
 *
 * **The example is loaded by Node rather than by a compiler.** Its TypeScript imports nothing by a `.js`
 * specifier and imports no type as a value, so type stripping runs it as it stands. That matters beyond
 * convenience: `tsx` resolves a bare specifier to its own instance of a package, so a driver that instruments
 * `@openai/agents` and then loads the example through `tsx` instruments a copy the example never uses, and the
 * run produces no span at all while exiting zero.
 *
 * Spans go wherever `OTEL_EXPORTER_OTLP_ENDPOINT` points, which is what `orchescope trace` sets before it runs
 * a command:
 *
 *   orchescope --cwd corpus/.cache/openai-agents-js-exercised trace -- \
 *     node corpus/runs/openai-agents-js/exercise.mjs <checkout> <node_modules>
 *
 * The environment is built by scripts/corpus/exercise.mjs.
 */

import { registerHooks } from 'node:module';
import { join, resolve } from 'node:path';
import { Readable } from 'node:stream';
import { pathToFileURL } from 'node:url';

const [, , checkoutArgument, modulesArgument] = process.argv;
const checkout = resolve(checkoutArgument ?? process.cwd());
const modules = resolve(modulesArgument ?? join(checkout, 'node_modules'));

/*
 * A position in the tree rather than a module. A bare specifier written in this file resolves the way the same
 * specifier written inside the checkout does, so the driver instruments the SDK instance the example runs.
 */
const insideTheEnvironment = pathToFileURL(join(modules, 'orchescope-corpus-driver.mjs')).href;
const isBare = (specifier) =>
  !specifier.startsWith('.') && !specifier.startsWith('/') && !specifier.includes(':');

registerHooks({
  resolve: (specifier, context, nextResolve) =>
    context.parentURL === import.meta.url && isBare(specifier)
      ? nextResolve(specifier, { ...context, parentURL: insideTheEnvironment })
      : nextResolve(specifier, context),
});

const { OTLPTraceExporter } = await import('@opentelemetry/exporter-trace-otlp-http');
const { BatchSpanProcessor, NodeTracerProvider } = await import('@opentelemetry/sdk-trace-node');
const { OpenAIAgentsInstrumentation } = await import(
  '@arizeai/openinference-instrumentation-openai-agents'
);
const agents = await import('@openai/agents');

const provider = new NodeTracerProvider({
  spanProcessors: [new BatchSpanProcessor(new OTLPTraceExporter())],
});
provider.register();
new OpenAIAgentsInstrumentation({ tracerProvider: provider }).manuallyInstrument(agents);

/** The transfer and the tool this conversation is meant to walk, in the order it walks them. */
const SCRIPT = [
  ['transfer_to_Seat_Booking_Agent', {}],
  ['update_seat', { confirmationNumber: 'IR-D204', seatNumber: '14A' }],
];

const used = new Set();

const functionCall = (name, args) => ({
  type: 'function_call',
  callId: `corpus-${used.size + 1}`,
  name,
  arguments: JSON.stringify(args),
  status: 'completed',
});

const finalAnswer = (text) => ({
  type: 'message',
  role: 'assistant',
  status: 'completed',
  content: [{ type: 'output_text', text }],
});

const scriptedModel = {
  getResponse(request) {
    const offered = (name) =>
      (request.handoffs ?? []).some((handoff) => handoff.toolName === name) ||
      (request.tools ?? []).some((tool) => tool.name === name);
    for (const [name, args] of SCRIPT) {
      if (!used.has(name) && offered(name)) {
        const call = functionCall(name, args);
        used.add(name);
        return Promise.resolve({ usage: new agents.Usage(), output: [call] });
      }
    }
    return Promise.resolve({
      usage: new agents.Usage(),
      output: [finalAnswer('Your seat is now 14A.')],
    });
  },
  /*
   * The example never streams. Answering a request that was never made would be a fiction, so this refuses
   * rather than returning nothing, and a checkout that starts streaming fails here instead of going quiet.
   */
  getStreamedResponse() {
    throw new Error('the customer service example does not stream, and this model does not either');
  },
};

agents.setDefaultModelProvider({ getModel: async () => scriptedModel });

const conversation = Readable.from([
  'I want to change my seat to 14A. My confirmation number is IR-D204.\n',
]);
Object.defineProperty(process, 'stdin', { value: conversation, configurable: true });

await import(pathToFileURL(join(checkout, 'examples/customer-service/index.ts')).href);

// The example's loop ends at end of file, and its turn is still in flight when the import resolves.
await new Promise((settle) => process.once('beforeExit', settle));

// What the run did, so a corpus log says whether the transfer was reached rather than only that it exited.
process.stdout.write(`script used: ${[...used].join(', ')}\n`);
await provider.shutdown();

File: corpus/runs/openai-cs-agents-demo/exercise.py
"""Exercise the airline agents of the pinned customer service demo so a real run can be joined to the declared graph.

The two entries that carry a run before this one are hermetic: each uses its library's own offline model, so neither
needs a credential and neither costs anything. That is the right property for a corpus entry and it is also the reason
the join has only ever been measured against a model that answers from a schema. This entry is the other case, an
application whose agents hand off to one another, and the handoff is the thing no offline model has exercised.

The OpenAI Agents SDK carries its own tracing, which exports to OpenAI's platform rather than over OTLP, so the spans
this join needs come from the OpenInference instrumentor instead. It emits `openinference.span.kind`, `llm.model_name`
and `tool.name`, which is what `packages/traces/src/attributes.ts` already decodes into `chat`, `embeddings` and
`execute_tool`.

Spans go wherever `OTEL_EXPORTER_OTLP_ENDPOINT` points, which is what `orchescope trace` sets before it runs a command:

    orchescope --cwd corpus/.cache/openai-cs-agents-demo trace -- \
      corpus/.cache/venvs/openai-cs-agents-demo/bin/python corpus/runs/openai-cs-agents-demo/exercise.py

`OPENAI_API_KEY` has to be in the environment and the run reaches the provider, which is what makes this entry
different from the other two and is why it is not part of `--exercise` by default.
"""

from __future__ import annotations

import asyncio
import sys
from pathlib import Path

from openinference.instrumentation.openai_agents import OpenAIAgentsInstrumentor
from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

CHECKOUT = Path(sys.argv[1]).resolve() if len(sys.argv) > 1 else Path.cwd()
sys.path.insert(0, str(CHECKOUT / "python-backend"))

provider = TracerProvider()
provider.add_span_processor(BatchSpanProcessor(OTLPSpanExporter()))
trace.set_tracer_provider(provider)

# Instrumented before the SDK is imported, so every agent and tool call is wrapped rather than only the later ones.
OpenAIAgentsInstrumentor().instrument(tracer_provider=provider)

from datetime import datetime  # noqa: E402

from agents import Runner  # noqa: E402
from chatkit.types import ThreadMetadata  # noqa: E402

from airline.agents import triage_agent  # noqa: E402
from airline.context import AirlineAgentChatContext, create_initial_context  # noqa: E402
from airline.demo_data import apply_itinerary_defaults  # noqa: E402
from memory_store import MemoryStore  # noqa: E402


def chat_context() -> AirlineAgentChatContext:
    """The context the demo's own server builds, assembled without the server.

    `AirlineAgentChatContext` is a ChatKit `AgentContext`, so it requires a thread, a store and a request context
    before any agent can run. The demo ships `MemoryStore` for exactly this and its server seeds the state with a
    demo itinerary, which the tools then read; an empty state reaches a tool that has nothing to answer from.
    """
    store = MemoryStore()
    state = create_initial_context()
    apply_itinerary_defaults(state)
    return AirlineAgentChatContext(
        thread=ThreadMetadata(id=store.generate_thread_id({}), created_at=datetime.now()),
        store=store,
        request_context={},
        state=state,
    )


async def main() -> None:
    """One question the triage agent has to hand off to answer.

    A seat change is chosen deliberately over a question triage can answer alone: the declared graph draws handoffs
    out of triage, and a run that never takes one joins to the agent and says nothing about the edges.
    """
    result = await Runner.run(
        triage_agent,
        "I want to change my seat to 14A.",
        context=chat_context(),
    )
    print(result.final_output)


if __name__ == "__main__":
    asyncio.run(main())
    provider.force_flush()
    provider.shutdown()
